File: src/components/HexMap/utils/pathfinding/pathfinding.ts
import {getNeighbors, type Hex} from "./getNeighbors";


function hexDistance(a: Hex, b: Hex): number {
    const ax = a.col;
    const az = a.row - (a.col - (a.col & 1)) / 2;
    const ay = -ax - az;
    const bx = b.col;
    const bz = b.row - (b.col - (b.col & 1)) / 2;
    const by = -bx - bz;
    return Math.max(Math.abs(ax - bx), Math.abs(ay - by), Math.abs(az - bz));
}


export function getReachableByPath(
    start: Hex,
    range: number,
    blocked: Set<string>
): Hex[] {
    const visited = new Set<string>();
    const queue: { hex: Hex; dist: number }[] = [{hex: start, dist: 0}];
    const result: Hex[] = [];

    while (queue.length > 0) {
        const {hex, dist} = queue.shift()!;
        const key = `${hex.col},${hex.row}`;
        if (visited.has(key)) continue;
        visited.add(key);
        result.push(hex);

        if (dist >= range) continue;

        for (const n of getNeighbors(hex)) {
            const nKey = `${n.col},${n.row}`;
            if (!blocked.has(nKey) && !visited.has(nKey)) {
                queue.push({hex: n, dist: dist + 1});
            }
        }
    }

    return result;
}


export function findPath(start: Hex, end: Hex, allowed: Hex[]): Hex[] {
    const allowedSet = new Set(allowed.map((h) => `${h.col},${h.row}`));
    const open: Hex[] = [start];
    const cameFrom = new Map<string, string>();
    const gScore = new Map<string, number>();
    const fScore = new Map<string, number>();

    const startKey = `${start.col},${start.row}`;
    gScore.set(startKey, 0);
    fScore.set(startKey, hexDistance(start, end));

    while (open.length > 0) {
        const current = open.reduce((a, b) =>
            (fScore.get(`${a.col},${a.row}`) ?? Infinity) <
            (fScore.get(`${b.col},${b.row}`) ?? Infinity)
                ? a
                : b
        );

        if (current.col === end.col && current.row === end.row) {
            return reconstruct(cameFrom, current);
        }

        open.splice(open.indexOf(current), 1);
        const currentKey = `${current.col},${current.row}`;

        for (const n of getNeighbors(current)) {
            const key = `${n.col},${n.row}`;
            if (!allowedSet.has(key)) continue;

            const tentativeG = (gScore.get(currentKey) ?? Infinity) + 1;
            if (tentativeG < (gScore.get(key) ?? Infinity)) {
                cameFrom.set(key, currentKey);
                gScore.set(key, tentativeG);
                fScore.set(key, tentativeG + hexDistance(n, end));
                if (!open.some((x) => x.col === n.col && x.row === n.row)) {
                    open.push(n);
                }
            }
        }
    }

    return [];
}


function reconstruct(cameFrom: Map<string, string>, current: Hex): Hex[] {
    const totalPath = [current];
    let currentKey = `${current.col},${current.row}`;
    while (cameFrom.has(currentKey)) {
        const prev = cameFrom.get(currentKey)!;
        const [col, row] = prev.split(",").map(Number);
        totalPath.unshift({col, row});
        currentKey = prev;
    }
    return totalPath;
}

File: backend/app/game/pathfinding.py
from collections import deque

from app.game.board import is_valid_tile


def hex_distance(a: dict[str, int], b: dict[str, int]) -> int:
    ax = a["col"]
    az = a["row"] - (a["col"] - (a["col"] & 1)) / 2
    ay = -ax - az
    bx = b["col"]
    bz = b["row"] - (b["col"] - (b["col"] & 1)) / 2
    by = -bx - bz
    return int(max(abs(ax - bx), abs(ay - by), abs(az - bz)))


def get_neighbors(hex_: dict[str, int]) -> list[dict[str, int]]:
    col = hex_["col"]
    row = hex_["row"]
    is_odd = col % 2 != 0
    directions = (
        [
            {"col": col + 1, "row": row},
            {"col": col + 1, "row": row + 1},
            {"col": col, "row": row + 1},
            {"col": col - 1, "row": row + 1},
            {"col": col - 1, "row": row},
            {"col": col, "row": row - 1},
        ]
        if is_odd
        else [
            {"col": col + 1, "row": row - 1},
            {"col": col + 1, "row": row},
            {"col": col, "row": row + 1},
            {"col": col - 1, "row": row},
            {"col": col - 1, "row": row - 1},
            {"col": col, "row": row - 1},
        ]
    )
    return [neighbor for neighbor in directions if is_valid_tile(neighbor["col"], neighbor["row"])]


def get_reachable_by_path(
    start: dict[str, int],
    movement_range: int,
    blocked: set[str],
) -> list[dict[str, int]]:
    visited: set[str] = set()
    queue = deque([(start, 0)])
    result: list[dict[str, int]] = []

    while queue:
        hex_, dist = queue.popleft()
        key = f'{hex_["col"]},{hex_["row"]}'
        if key in visited:
            continue
        visited.add(key)
        result.append(hex_)

        if dist >= movement_range:
            continue

        for neighbor in get_neighbors(hex_):
            neighbor_key = f'{neighbor["col"]},{neighbor["row"]}'
            if neighbor_key not in blocked and neighbor_key not in visited:
                queue.append((neighbor, dist + 1))

    return result


def find_path(
    start: dict[str, int],
    end: dict[str, int],
    allowed: list[dict[str, int]],
) -> list[dict[str, int]]:
    allowed_set = {f'{hex_["col"]},{hex_["row"]}' for hex_ in allowed}
    open_nodes = [start]
    came_from: dict[str, str] = {}
    g_score = {f'{start["col"]},{start["row"]}': 0}
    f_score = {f'{start["col"]},{start["row"]}': hex_distance(start, end)}

    while open_nodes:
        current = min(
            open_nodes,
            key=lambda node: f_score.get(f'{node["col"]},{node["row"]}', float("inf")),
        )

        if current["col"] == end["col"] and current["row"] == end["row"]:
            return _reconstruct_path(came_from, current)

        open_nodes.remove(current)
        current_key = f'{current["col"]},{current["row"]}'

        for neighbor in get_neighbors(current):
            key = f'{neighbor["col"]},{neighbor["row"]}'
            if key not in allowed_set:
                continue

            tentative_g = g_score.get(current_key, float("inf")) + 1
            if tentative_g < g_score.get(key, float("inf")):
                came_from[key] = current_key
                g_score[key] = tentative_g
                f_score[key] = tentative_g + hex_distance(neighbor, end)
                if neighbor not in open_nodes:
                    open_nodes.append(neighbor)

    return []


def _reconstruct_path(came_from: dict[str, str], current: dict[str, int]) -> list[dict[str, int]]:
    path = [current]
    current_key = f'{current["col"]},{current["row"]}'
    while current_key in came_from:
        prev = came_from[current_key]
        col, row = prev.split(",")
        path.insert(0, {"col": int(col), "row": int(row)})
        current_key = prev
    return path
